### PR TITLE
Fix #313: Configurable auth methods

### LINKED ISSUE #313: Configure the list of available Authentication mecanisms to activate.
In the Kinto Writer we want to Authenticate with LDAP, we should be able to rename Basic Auth in LDAP and hide the FxA authentication method.

### PR DESCRIPTION
**Pull request by leplatrem**
* [x] Basic refactor / implem
* [x] Fix problem ldap submits Authorization as undefined (ref #326)
* [x] Add anonymous method
* [ ] LDAP should return to login form if credentials are invalid (i.e. if user obtained in server info is undefined or does not start with `ldap:`

### CODE REVIEW

**Comment by leplatrem** [comment]
@n1k0 r?

File: src/reducers/settings.js [hunk 1]
```diff
@@ -5,9 +5,10 @@ import type { SettingsState } from "../types";
 const INITIAL_STATE: SettingsState = {
   maxPerPage: 200,
   singleServer: null,
+  authMethods: ["basicauth", "fxa", "ldap"]
 };
 
 // This is basically a noop as we don't support settings update at runtime (yet).
 export default function settings(state: SettingsState = INITIAL_STATE): SettingsState {
-  return state;
+  return {...INITIAL_STATE, ...state};
```
Comment (n1k0) [inline src/reducers/settings.js]: Well, that does the same thing no? :) I mean, no action is ever processed here, and so state never changes...

File: src/types.js [hunk 2]
```diff
@@ -280,7 +280,8 @@ export type ServerInfo = {
 
 export type SettingsState = {
   maxPerPage: number,
-  singleServer: ?string
+  singleServer: ?string,
+  authMethods: string[]
```
Comment (n1k0) [inline src/types.js]: We could create an `AuthMethod` type which would be `"basicauth" | "fxa" | "ldap"`, and here have `AuthMethod[]`

File: src/client.js [hunk 3]
```diff
@@ -13,7 +13,9 @@ function getAuthHeader(auth: AuthData): ?string {
       const {token}: {token: string} = auth.credentials;
       return "Bearer " + token;
     }
-    case "basicauth": {
+    // case "ldap": {
+    // case "basicauth":
```
Comment (n1k0) [inline src/client.js]: Nit: except if I'm missing something, I'd leave these uncommented

File: src/components/AuthForm.js [hunk 4]
```diff
@@ -218,32 +259,26 @@ export default class AuthForm extends Component {
 
   constructor(props: Object) {
     super(props);
+    const {settings: {authMethods}} = this.props;
+    const defaultAuth = authMethods[0];
```
Comment (n1k0) [inline src/components/AuthForm.js]: how about being explicit with just `"basicauth"`?

File: src/reducers/settings.js (hunk 1 above)
Comment (leplatrem) [inline src/reducers/settings.js]: I did that to allow to have default settings (ie. not specified in `KintoAdmin` component)

File: src/components/AuthForm.js (hunk 4 above)
Comment (leplatrem) [inline src/components/AuthForm.js]: But if `"basicauth"` is not in the list?

Comment (leplatrem) [inline src/components/AuthForm.js]: yes, since we make it configurable. And for example we could have a stack where only fxa is allowed

**Pull request merged**